src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,278 @@
+            {
+                return null;
+            }
+
+            return GetSniFromSslHandshake(sslHandshake);
+        }
+
+        // https://tools.ietf.org/html/rfc6101#section-5.6
+        // Handshake structure:
+        //   - HandshakeType msg_type (1 bytes) => 0x01 is client_hello
+        //   - uint24 length
+        //   - <msg_type> body
+        private static string GetSniFromSslHandshake(ReadOnlySpan<byte> sslHandshake)
+        {
+            // If not client hello then skip
+            if (sslHandshake.Length < 4 || sslHandshake[0] != 0x01)
+            {
+                return null;
+            }
+
+            int clientHelloLength = ReadUint24(sslHandshake.Slice(1));
+            ReadOnlySpan<byte> clientHello = sslHandshake.Slice(4);
+
+            if (clientHello.Length != clientHelloLength)
+            {
+                return null;
+            }
+
+            return GetSniFromClientHello(clientHello);
+        }
+
+        // 5.6.1.2. https://tools.ietf.org/html/rfc6101#section-5.6.1 - describes basic structure
+        // 2.1. https://www.ietf.org/rfc/rfc3546.txt - describes extended structure
+        // ClientHello structure:
+        //   - ProtocolVersion client_version (2 bytes)
+        //   - Random random (32 bytes => 4 bytes GMT unix timestamp + 28 bytes of random bytes)
+        //   - SessionID session_id (opaque type of max size 32 => size fits in 1 byte)
+        //   - CipherSuite cipher_suites (opaque type of max size 2^16-1 => size fits in 2 bytes)
+        //   - CompressionMethod compression_methods (opaque type of max size 2^8-1 => size fits in 1 byte)
+        //   - Extension client_hello_extension_list (opaque type of max size 2^16-1 => size fits in 2 bytes)
+        private static string GetSniFromClientHello(ReadOnlySpan<byte> clientHello)
+        {
+            // Skip ProtocolVersion and Random
+            ReadOnlySpan<byte> p = SkipBytes(clientHello, 34);
+
+            // Skip SessionID
+            p = SkipOpaqueType1(p);
+
+            // Skip cipher suites
+            p = SkipOpaqueType2(p);
+
+            // Skip compression methods
+            p = SkipOpaqueType1(p);
+
+            // is invalid structure or no extensions?
+            if (p.IsEmpty)
+            {
+                return null;
+            }
+
+            int extensionListLength = ReadUint16(p);
+            p = SkipBytes(p, 2);
+
+            if (extensionListLength != p.Length)
+            {
+                return null;
+            }
+
+            while (!p.IsEmpty)
+            {
+                string sni = GetSniFromExtension(p, out p);
+                if (sni != null)
+                {
+                    return sni;
+                }
+            }
+
+            return null;
+        }
+
+        // 2.3. https://www.ietf.org/rfc/rfc3546.txt
+        // Extension structure:
+        //   - ExtensionType extension_type (2 bytes) => 0x00 is server_name
```

Comment:
ExtensionType is a ushort enum.  You should define the enum, at least for the values you are using.